### DISCUSSION TITLE
How To represent the following flow

### DISCUSSION THREAD

**Discussion by ibrahimkteish** [pointfreeco/swift-composable-architecture#895] [Q&A]
Hello All, 

In my project I start using a back-port (iOS 10+) version of the composable architecture, I have a  question on how to design the following flow.

<img width="555" alt="Screenshot 2021-11-16 at 09 27 23" src="https://user-images.githubusercontent.com/2681868/141949245-cd80762a-4da1-40a6-a5ea-abbe68d28150.png">

`Flow B` and `C` are mutual exclusive so for them I represented the state as Enum 

```swift
enum SubFlow: Equatable {
case flowB(FlowBStructState)
case flowC(FlowCStructState)
}
//....

let subFlowReducer = .init { state, action, env in ... }
```

every thing working as expected. Now I want to continue refactoring the code to create a flow reducer for all 3 flows. The question is should `FlowA` be another case in my `enum` or a `struct` holding a optional property of type `SubFlow` as the following

```swift
struct Flow {
var prop1: String,
var prop2: Int
var subFlow: SubFlow?
}
```

I would appreciate if someone can help me design this in the best way.

**Comment by surajbarkale** [comment]
This is how I would model it:
```swift

enum SubFlow: Equatable {
  case flowB(FlowBState)
  case flowC(FlowCState)
}

struct FlowState: Equatable {
  var flowA: FlowAState
  var subFlow: SubFlow?
}

let flowReducer = Reducer<FlowState, FlowAction, FlowEnvironment>.combine(
  flowAReducer.pullback(\.flowA, /FlowAction.flowA, \.flowA),
  flowBReducer
    .pullback(/SubFlow.flowB, \.self, \.self)
    .pullback(\.subFlow, /FlowAction.flowB, \.flowB),
  flowCReducer
    .pullback(/SubFlow.flowC, \.self, \.self)
    .pullback(\.subFlow, /FlowAction.flowC, \.flowC),
  Reducer { state, action, environment in
    ...
  }
)
```
